src/features/machine/rectMachine.js

```diff
@@ -0,0 +1,333 @@
+import Victor from 'victor'
+import Machine from './machine'
+import { distance } from '../../common/Geometry'
+
+export default class RectMachine extends Machine {
+  constructor(vertices, settings) {
+    super()
+    this.vertices = vertices
+    this.settings = settings
+    this.sizeX = Math.abs((settings.maxX - settings.minX) / 2.0)
+    this.sizeY = Math.abs((settings.maxY - settings.minY) / 2.0)
+  }
+
+  addEndpoints() {
+    if (this.settings.rectOrigin.length === 1) {
+      // OK, let's assign corners indices:
+      // [1]   [2]
+      //
+      //
+      // [0]   [3]
+      const corner = this.settings.rectOrigin[0]
+      const corners = [
+        new Victor(-this.sizeX, -this.sizeY),
+        new Victor(-this.sizeX, this.sizeY),
+        new Victor(this.sizeX, this.sizeY),
+        new Victor(this.sizeX, -this.sizeY)
+      ]
+
+      const first = this.vertices[0]
+      const last = this.vertices[this.vertices.length-1]
+      const maxRadius = Math.sqrt(Math.pow(2.0*this.sizeX, 2.0) + Math.pow(2.0*this.sizeY, 2.0)) / 2.0
+      let scale, outPoint
+
+      if (first.magnitude() <= last.magnitude()) {
+        // It's going outward
+        scale = maxRadius / last.magnitude()
+        outPoint = last
+      } else {
+        scale = maxRadius / first.magnitude()
+        outPoint = first
+      }
+
+      let clipped = this.clipLine(
+        this.vertices[this.vertices.length - 1],
```

Comment:
We just figured out if we were growing or shrinking, and used that to find the ideal perimeter point, but then we always add this to the last point. I think adding it to the last point makes the most sense. So I'm not sure why I was checking for growing or shrinking.

Reply:
It looks funny if I always add it to the last point, instead of the outermost point. So I changed this around to insert the points to the inside if the pattern is shrinking.